[core.date] migrate Calendar to ZonedDateTime and add some new functionality

CrazyIvan359 opened the pull request:
Migrate to using ZonedDateTime from java.time, joda and java.util datetimes are depreciated. Added timezone support. Added simpler formatter with default matching OH default formatting. Added days, hours, minutes, seconds between functions. Added `__all__` import list to limit overlap when using wildcard imports on this module

openhab-5iver commented:
Thank you for the contribution! I have to admit that other than testing, I haven't used this module. Could you list the previous functionality that is no longer present, if any? And do you have a test script that you used, so that we can include it in the repo? This would also save me some time in testing.

You have 'aliases for compatibility', but I'd rather this be reversed, and use the short names as the aliases. Also, use more readable names for the other functions and variables. Sorry for this... I've been called out for readability issues too!

The documentation will need to be updated with your changes. Scratch that... there is none! I'll leave it up to you, if you'd like to get it added. If not, I'll add it in.

CrazyIvan359 commented:
No problem! I've been working on migrating to JSR223 over the last couple of weeks since I came across discussion of removing Xtend entirely. Decided I would use python since I know a fair bit of it already and JS is still fairly new to me (not counting my Xtend experience with OH). Definitely got the impression you never used this module much, before I even started using it I decided to do this rewrite. 

The only thing that would be a breaking change is your old formatter that you just brought in from `java.util`. Your old version changed everything to `calendar` and then to the requested format, where as mine changes to `ZonedDateTime` and then to the requested format, this changes nothing in userland though. I will get the naming switched to your style and use my short names as aliases.

Not to worry about nitpicking for readability,it's not helpful if you can't understand how to use it. I am only a programmer by hobby, so could you clear a few things up as to what you need me to do:
- can you give me an example of one of the variables and how you would name it?
- I didn't use a test script, I just used it in my production OH and debugged that way. What should this test script do?

CrazyIvan359 commented:
Oh, something I thought I would mention for use in this entire project is the `__all__` property. I discovered it while digging around for python things to use in my rules. If the `__all__` property is present in the file and you import it using `from [file] import *` only the things in the list `__all__` will be import, instead of the entire namespace of the file (including anything you've imported). It seems prudent to use this given the novice tendency to use wildcard instead of specific imports, so as not to dirty the namespace in the rule file too much.

I could go through and add these if you're busy with other things

openhab-5iver commented:
> and JS is still fairly new to me

It is also only ES 5.1 in JDK8, which is really limited. JDK9 will get to ES6, and have classes, which will allow the JS libraries to match the Jython functionality.

> The only thing that would be a breaking change is your old formatter that you just brought in from `java.util`.

Whoa hold up a sec. We inherited most of this repo from a developer that parted ways with OH. I've only stepped in to help maintain it. I've added a few things, but it feels weird when people think of me as the owner of the code. There are 2 other maintainers that have helped out in the past, and will hopefully come out of the shadows once I get the lucid repo migrated back in to this one. If someone is active and interested, we'd bring them in as a maintainer too. I'm getting spread pretty thin right now, trying to get scripted Actions into the NGRE (this is basically done, if they'd stop breaking my IDE with the ESH migration), making bundles for Jython and Groovy, working on ways to incorporate the old rule format into the NGRE or JSR223, extending the automation API, etc. So, help like this is **very** much appreciated! Anyhow...

If stripping Calendar is the only breaking change, then I think we will be fine. I'll keep an eye out for it in the lucid code.

> I am only a programmer by hobby, so could you clear a few things up as to what you need me to do:

    * can you give me an example of one of the variables and how you would name it?

    * I didn't use a test script, I just used it in my production OH and debugged that way. What should this test script do?

zdt, c, dt, min,milli, etc. It takes a few more keys, but much more readable when things are fully typed out. I've also done my best to only use camel case for classes and underscores for the rest. Someday we may have some coding guidelines. :wink: For the test script, I just put together scripts to test all of the functionality that I implement and log the results. There's also the core.testing module that can be useful. I wouldn't worry too much if you don't have anything, but it could save some time if you do.

> If the `__all__` property is present in the file and you import it using `from [file] import *` only the things in the list `__all__` will be imported, instead of the entire namespace of the file (including anything you've imported in it). It seems prudent to use this given the novice tendency to use wildcard instead of specific imports, so as not to dirty the namespace in the rule file too much.

I've debated this a few times, and think I have it somewhere in a backlog, but ultimately always decided that it wasn't worth the effort. The modules are pretty clean, and there is not much (any?) namespace conflict. It's also best practice to not use import *. If you feel really strongly about it, then go for it, but IMO, it just adds to the effort. Realize though that these helper libraries are only temporary until the API(s) are built out and make most of them obsolete.

CrazyIvan359 commented:
> Whoa hold up a sec. We inherited most of this repo from a developer that parted ways with OH. I've only stepped in to help maintain it. I've added a few things, but it feels weird when people think of me as the owner of the code. 

Sorry, that wasn't meant to be accusatory, I meant only to specify that it was existing code not something I had written. These helper libraries have made my ongoing transition to JSR223 rather painless! I would have ended up writing some kind of wrapper of my own, but it wouldn't have been anywhere near as elegant as this.


> If stripping Calendar is the only breaking change, then I think we will be fine. I'll keep an eye out for it in the lucid code.

Calendar is staying, it is depreciated in JDK8 and up I believe, but remains available for now. I removed the `date_formatter` that was just an import of the `java.util.SimpleDateFormat` with the same format as OH but less the timezone. I found that this was not easy to use or evident how to, so I made a function that returns the date as a string in the passed format, with the OH format as the default.


> zdt, c, dt, min,milli, etc. It takes a few more keys, but much more readable when things are fully typed out. I've also done my best to only use camel case for classes and underscores for the rest. Someday we may have some coding guidelines. 😉

I will make these changes. This is what I thought you were talking about, but at that point I was also under the impression you had written this file yourself. Most of those variables are the ones from the original version.


> I've debated this a few times, and think I have it somewhere in a backlog, but ultimately always decided that it wasn't worth the effort. The modules are pretty clean, and there is not much (any?) namespace conflict. It's also best practice to not use import *. If you feel really strongly about it, then go for it, but IMO, it just adds to the effort.

I agree that `import *` should never be used, but I've seen plenty of guides by novices suggesting it be done (out of laziness, I presume). I have scoured any of the other files like I have this one, but I think the namespace it pretty clean. This file I think does merit the effort since there are so many similarly named date types being imported.


>For the test script, I just put together scripts to test all of the functionality that I implement and log the results. There's also the core.testing module that can be useful. I wouldn't worry too much if you don't have anything, but it could save some time if you do.

I will look into this and write something up. Might be this weekend before I have time though


>Realize though that these helper libraries are only temporary until the API(s) are built out and make most of them obsolete.

I wasn't aware such efforts were in progress. Will jython still be supported once they are complete?

CrazyIvan359 commented:
Ok, variable names clarified and camelCase removed, consistency is important!
I also remove some unneeded methods in the `pythonTimezone` class, I had written them when I was going to favour python datetime but changed my mind and went with ZonedDateTime instead.
And I added a more detailed description of the module and its use at the top.

Next change I have to work on it I will look into writing up a test script and maybe some kind of documentation

openhab-5iver commented:
I added a small fix for the java.sql.Date and java.sql.Timestamp workaround, and helped polish some of the text. I don't see any breaking changes, so I'll get this merged in, but let's add some documentation and tests.

CrazyIvan359 commented:
Awesome! My first time contributing on Github is merged! 

That sql workaround was from the original, I hadn't looked into it at all since it seemed to be working (or at least not breaking anything). 

I will see about writing up a doc for it tomorrow. Should I make a page entirely for this module?

If I have time I'll also work on a unittest for it. Feel free to tag me in any issues that get opened against it, I can look after them.

openhab-5iver commented:
> That sql workaround was from the original, I hadn't looked into it at all since it seemed to be working (or at least not breaking anything).

Yep. I had run into an issue with it a long while back. Forgot about it and never put in a commit.

> Should I make a page entirely for this module?

No need for a whole page... would be great if you could make an entry in here... https://github.com/OH-Jython-Scripters/openhab2-jython/blob/master/Docs/Jython-Modules.md.
